File: src/utils/sign.ts
import CryptoJS from 'crypto-js'
import { ecdsaSign } from 'secp256k1'
import _ from 'lodash'
import { RawKey } from '@terra-money/terra.js'
import { Base } from 'use-station/src'

export const createSignedTransaction = async (
  tx: Record<string, any>,
  privateKey: string,
  request: Base
): Promise<string> => {
  const key = new RawKey(Buffer.from(privateKey, 'hex'))
  const { sequence, account_number, chain_id } = request
  const _req = {
    sequence,
    accountNumber: account_number,
    chainId: chain_id,
  }
  const signMessage = createSignMessage(tx, _req)

  const signature = signWithPrivateKey(
    signMessage,
    Buffer.from(privateKey, 'hex')
  )

  const signatureObject = createSignature(
    signature,
    sequence,
    account_number,
    key.publicKey
  )

  return JSON.stringify(
    Object.assign({}, tx, { signatures: [signatureObject] })
  )
}

const signWithPrivateKey = (
  signMessage: string,
  privateKey: Uint8Array
): Buffer => {
  const signMessageString =
    typeof signMessage === 'string'
      ? signMessage
      : JSON.stringify(signMessage)
  const signHash = Buffer.from(
    CryptoJS.SHA256(signMessageString).toString(),
    `hex`
  )
  const { signature } = ecdsaSign(signHash, privateKey)

  return Buffer.from(signature)
}

const createSignature = (
  signature: Buffer,
  sequence: string,
  accountNumber: string,
  publicKey?: Buffer
): {
  signature: any
  account_number: any
  sequence: any
  pub_key: {
    type: string
    value: any
  }
} => ({
  signature: signature.toString('base64'),
  account_number: accountNumber,
  sequence,
  pub_key: {
    type: 'tendermint/PubKeySecp256k1',
    value: publicKey?.toString('base64'),
  },
})

const createSignMessage = (
  jsonTx: Record<string, any>,
  {
    sequence,
    accountNumber,
    chainId,
  }: {
    sequence: string
    accountNumber: string
    chainId: string
  }
): string => {
  const fee = {
    amount: jsonTx.fee.amount || [],
    gas: jsonTx.fee.gas,
  }

  return JSON.stringify(
    removeEmptyProperties({
      fee,
      memo: jsonTx.memo,
      msgs: jsonTx.msg,
      sequence,
      account_number: accountNumber,
      chain_id: chainId,
    })
  )
}

const removeEmptyProperties = (
  obj: Record<string, unknown>
): Record<string, unknown> => {
  return _.pickBy(obj, _.identity)
}

File: ios/TerraStation/TerraWalletLib/SDK/Sign.swift
import Foundation

class Sign {

    private var privateKey:Data
    private var publicKey:Data
    private var sequence:String
    private var accountNumber:String
    private var chainId:String

    init(hexPrivateKey:String, hexPublicKey:String, sequence:String, accountNumber:String, chainId:String) {

        self.privateKey = hexPrivateKey.hexToData
        self.publicKey = hexPublicKey.hexToData
        self.sequence = sequence
        self.accountNumber = accountNumber
        self.chainId = chainId
    }

    func sign(message:Dictionary<String, Any>) -> Dictionary<String, Any> {
      let signMessage = createSignMessage(json: message)
      //sort json by key.
      guard let sortedString = jsonSort(signMessage: signMessage) else {
          return [:]
      }

      print("signed : \(sortedString)")

      let signedMessage = signWithPrivateKey(message: sortedString)
      let signatureJson = createSignature(signature: signedMessage)
      let signedTx = createSignedTx(json: message, signature: signatureJson)
      let broadcastBody = createBroadcastBody(json: signedTx, returnType: "block")

      return broadcastBody
    }

    private func createSignMessage(json:Dictionary<String, Any>) -> Dictionary<String, Any> {

        let json_fee = json["fee"] as? Dictionary<String, Any>
        var fee = Dictionary<String, Any>()
        fee["amount"] = json_fee?["amount"] ?? []
        fee["gas"] = json_fee?["gas"]

        var message = Dictionary<String, Any>()
        message["fee"] = fee
        message["memo"] = json["memo"] ?? ""
        message["msgs"] = json["msg"]
        message["sequence"] = self.sequence
        message["account_number"] = self.accountNumber
        message["chain_id"] = self.chainId

        return message
    }

    private func jsonSort(signMessage:Dictionary<String, Any>) -> String? {
      guard let sorted = try? JSONSerialization.data(withJSONObject: signMessage, options: .sortedKeys),
        let sortedString = String(data: sorted, encoding: .utf8)?.replacingOccurrences(of: "\\/", with: "/") else {
          return nil
      }

      return sortedString
    }

    private func signWithPrivateKey(message:String) -> Data {
        guard let jsonData = message.data(using: .utf8) else { return Data() }
        let hashedMessage = Utils.sha256(data: jsonData)

        var curve = secp256k1
        var signature = Data(repeating: 0, count: 64)
        signature.withUnsafeMutableBytes { (sig) in
          ecdsa_sign_digest(&curve, self.privateKey.bytes, hashedMessage.bytes, sig, nil, nil)
        }

        return signature
    }

    private func createSignature(signature:Data) -> Dictionary<String, Any> {

        var dic = Dictionary<String, Any>()
        dic["signature"] = signature.base64EncodedString()
        dic["account_number"] = self.accountNumber
        dic["sequence"] = self.sequence
        dic["pub_key"] = ["type":"tendermint/PubKeySecp256k1",
                          "value":self.publicKey.base64EncodedString()]
        return dic
    }

    private func createSignedTx(json:Dictionary<String, Any>, signature:Dictionary<String, Any>) -> Dictionary<String, Any> {

        var data = json
        data["signatures"] = [signature]
        return data
    }

    private func createBroadcastBody(json:Dictionary<String, Any>, returnType:String) -> Dictionary<String, Any> {

        var data = Dictionary<String, Any>()
        data["tx"] = json
        data["mode"] = returnType
        return data
    }


}
